### PR TITLE
Golang try vendor=null, then set vendor hash

### PR DESCRIPTION
**Pull request by c00w**
This always tries to set vendor=null (which is preferable), before falling back to vendor=hash.

### CODE REVIEW

**Comment by c00w** [comment]
The tests pass with 
```
nix-shell --run "cabal v2-test"
```

However, running nix-build & the output, exits with
```
../nixpkgs-update/result/bin/nixpkgs-update update "vultr-cli 0.3.1 0.3.2"
``` on both master and this PR. I didn;t bother to set my github token, which might be it.

**Comment by c00w** [comment]
After setting up a github token, it now fails with complaints about getent :(

**Comment by c00w** [comment]
```
./result/bin/nixpkgs-update update "tflint 0.15.0 0.15.1"
exec: "getent": executable file not found in $PATH
nixpkgs-update: Received ExitFailure 1 when running
Raw command: hub clone nixpkgs /home/colin/.cache/nixpkgs
```

This is on master and my PR, so I think the guidance for setup is just insufficient in the README.

**Comment by bhipple** [comment]
You need `hub` in your `$PATH`: https://github.com/ryantm/nixpkgs-update#update-tutorial

**Comment by c00w** [comment]
Hmmm. hub is in my path (and manually running "hub clone nixpkgs /home/colin/.cache/nixpkgs" works.

Now that I've run it, I'm getting a new error with 
```
../nixpkgs-update/result/bin/nixpkgs-update update "vultr-cli 0.3.1 0.3.2"
nixpkgs-update: /home/colin/.cache/nixpkgs/.git/FETCH_HEAD: getModificationTime:getFileStatus: does not exist (No such file or directory)
```

**Comment by c00w** [comment]
@bhipple Pushing new version. vultr-cli seems like a good test candidate :) As well as something which falls back - i.e. gobetween

**Comment by c00w** [comment]
The following commands both pass for me

`./result/bin/nixpkgs-update update "minio-client 2020-04-25T00-43-23Z 2020-05-28T23-43-36Z" `

`./result/bin/nixpkgs-update update "vultr-cli 0.3.1 0.3.2"`

File: src/Rewrite.hs [hunk 1]
```diff
@@ -199,16 +200,19 @@ golangModuleVersion log args@(Args _ attrPth drvFile drvContents) = do
...
+      _ <- if isLeft ok
+      then do
+        _ <- liftIO $ T.writeFile drvFile original
+        _ <- lift $ File.replaceIO oldVendorSha256 Nix.sha256Zero drvFile
+        newVendorSha256 <- Nix.getHashFromBuild attrPth
+        _ <- lift $ File.replaceIO Nix.sha256Zero newVendorSha256 drvFile
+        lift . log $ "Replaced vendorSha256 with " <> newVendorSha256
+      else  do
+        lift . log $ "Set vendorSha256 to null"
+      -- Note that on some small bumps, this may not actually change if go.sum did not
```
Comment (bhipple) [inline src/Rewrite.hs:215]: This comment goes above line 212.

File: src/Rewrite.hs [hunk 2]
```diff
@@ -199,16 +200,19 @@ golangModuleVersion log args@(Args _ attrPth drvFile drvContents) = do
...
+      _ <- lift $ File.replaceIO ("\"" <> oldVendorSha256 <> "\"") "null" drvFile
+      ok <- runExceptT $ Nix.build attrPth
+      _ <- if isLeft ok
+      then do
+        _ <- liftIO $ T.writeFile drvFile original
+        _ <- lift $ File.replaceIO oldVendorSha256 Nix.sha256Zero drvFile
+        newVendorSha256 <- Nix.getHashFromBuild attrPth
+        _ <- lift $ File.replaceIO Nix.sha256Zero newVendorSha256 drvFile
+        lift . log $ "Replaced vendorSha256 with " <> newVendorSha256
+      else  do
```
Comment (bhipple) [inline src/Rewrite.hs:213]: Two spaces.  Can you run through `ormolu` to fix formatting?
https://github.com/ryantm/nixpkgs-update#development

**Comment by zowoq** [comment]
Probably need to try with `deleteVendor` removed as well. https://github.com/NixOS/nixpkgs/pull/89907

**Comment by c00w** [comment]
Updated - I'll go play with delete vendor, but I can open another PR, no need to block merging this on that.

**Comment by c00w** [comment]
@zowoq I added removel of deleteVendor support. Need to still add support for adding it (as well as runVend I guess).

**Comment by c00w** [comment]
@bhipple  Should be good for another review.

File: src/File.hs [hunk 3]
```diff
@@ -56,3 +57,23 @@ replaceIO find replacement file =
...
+  Member File r =>
+  Text ->
+  FilePath ->
+  Sem r Bool
+dropLine find file = do
+  contents <- File.read file
+  let newContents = T.unlines $ filter (not . T.isInfixOf find) (T.lines contents)
+  when (contents /= newContents) $ do
+    File.write file newContents
+  return $ contents /= newContents
```
Comment (bhipple) [inline src/File.hs:71]: This looks fragile.   Can we do better?  Maybe ask `nix eval` for line info, or implement it as `dropAttrLine` that deletes `^\s*$attr.*=.*;`?  Also could use some test cases, which we're trying to improve coverage of.

Everything you had up until the `deleteVendor` stuff looks good; if you rebase to just have that I can merge those portions and we can handle this in another PR (up to you).

**Comment by zowoq** [comment]
> I added removel of deleteVendor support.

Thanks!

> Need to still add support for adding it (as well as runVend I guess).

I think these (hopefully uncommon) cases could be dealt with manually as they should appear as failures in the bot logs, my concern was once they are added to a package they might not be noticed during review and removed.

**Comment by c00w** [comment]
@bhipple Dropped deleteVendor :)

**Pull request merged**